elastic_service/elastic_handler.go

```diff
@@ -0,0 +1,167 @@
+	errors "github.com/fiverr/go_errors"
+	"github.com/olivere/elastic/v7"
+	"strconv"
+	"strings"
+)
+
+var _ datastore.BookStorer = &ElasticHandler{}
+
+type ElasticHandler struct {
+	Client *elastic.Client
+}
+
+func NewElasticHandler(client *elastic.Client) ElasticHandler {
+	return ElasticHandler{
+		client,
+	}
+}
+
+func (e *ElasticHandler) UpdateBook(ctx context.Context, title string, id string) (err error) {
+	_, err = e.Client.Update().
+		Index(consts.Index).
+		Id(id).
+		Doc(map[string]interface{}{elastic_fields.Title: title}).
+		Do(ctx)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("couldn't update book with title %s and id %s", title, id))
+	}
+	return nil
+}
+
+func (e *ElasticHandler) InsertBook(ctx context.Context, body []byte) (string, error) {
+	res, err := e.Client.Index().
+		Index(consts.Index).
+		BodyString(string(body)).
+		Do(ctx)
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("couldn't update book with body: %s", string(body)))
+	}
+
+	return res.Id, nil
+}
+func (e *ElasticHandler) GetBook(ctx context.Context, id string) (src *models.Book, err error) {
+	get, err := e.Client.Get().
+		Index(consts.Index).
+		Id(id).
+		Do(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("couldn't get book form server with id: %s", id))
+	}
+
+	var book models.Book
+	if err = json.Unmarshal(get.Source, &book); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("couldn't unmarshal result of book with id: %s, in getBoook function ", id))
+	}
+	return &book, nil
+}
+
+func (e *ElasticHandler) DeleteBook(ctx context.Context, id string) error {
+	_, err := e.Client.Delete().
+		Index(consts.Index).
+		Id(id).
+		Do(ctx)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("couldn't delete book with id: %s", id))
+	}
+	return nil
+}
+
+func buildSearchQuery(fields map[string]string) (s *elastic.BoolQuery, err error) {
+	var from, to int
+	if price, ok := fields["price_range"]; ok && price != "" {
+		priceSplit := strings.Split(string(fields["price_range"]), "-")
+		if len(priceSplit) != 2 {
+			return nil, errors.New("failed to pars `price_range` field")
+		}
+		if from, err = strconv.Atoi(priceSplit[0]); err != nil {
+			return nil, errors.New("failed to pars `price_range` field")
+		}
+		if to, err = strconv.Atoi(priceSplit[1]); err != nil {
+			return nil, errors.New("failed to pars `price_range` field")
+		}
+	}
+
+	if from < consts.MinPrice {
+		return nil, errors.New(fmt.Sprintf("illegal price range, price should be higher than %d", consts.MinPrice))
+	}
+	if to > consts.MaxPrice {
+		return nil, errors.New(fmt.Sprintf("illegal price range, price should be lower than %d", consts.MaxPrice))
+	}
+
+	q := elastic.NewBoolQuery()
+
+	if fields["title"] != "" {
+		q.Must(elastic.NewMatchQuery(elastic_fields.Title, fields["title"]))
+	}
+	if fields["author"] != "" {
+		q.Must(elastic.NewMatchQuery(elastic_fields.Author, fields["author"]))
+	}
+	if to != 0 {
```

Comment:
If fields["price_range"] do not exist, It is best not to add a range query to elastic.

Reply:
in that case we do not enter the if statement (because to will be zero - default value of int), so we won't be adding the range query

Reply:
Got it, I think it is more intuitive to ask regarding the query param...